wire-library/wire-runtime-swift/src/main/swift/ProtoEnum.swift

```diff
@@ -0,0 +1,148 @@
+
+        if let string = try? container.decode(String.self) {
+            guard let value = Self(string) else {
+                throw ProtoDecoder.Error.unknownEnumString(type: Self.self, string: string)
+            }
+            self = value
+        } else {
+            // If the value wasn't a string, then look for the field index instead.
+            let fieldNumber = try container.decode(UInt32.self)
+            guard let value = Self(rawValue: fieldNumber) else {
+                throw ProtoDecoder.Error.unknownEnumCase(type: Self.self, fieldNumber: fieldNumber)
+            }
+
+            self = value
+        }
+    }
+
+    public func encode(to encoder: Encoder) throws {
+        var container = encoder.singleValueContainer()
+
+        switch encoder.enumEncodingStrategy {
+        case .string:
+            try container.encode(description)
+
+        case .integer:
+            try container.encode(rawValue)
+        }
+    }
+}
+
+/// This  type gives me access to `decoder.enumDecodingStrategy`
+private struct BoxedEnum<T: ProtoEnum> : Decodable {
+    let value: T?
+
+    init(value: T? = nil) {
+        self.value = value
+    }
+
+    init(from decoder: Decoder) throws {
+        let container = try decoder.singleValueContainer()
+
+        guard !container.decodeNil() else {
+            self.init()
+            return
+        }
+
+        switch decoder.enumDecodingStrategy {
+        case .returnNil:
+            let value = try? container.decode(T.self)
+            self.init(value: value)
+
+        case .throwError:
+            let value = try container.decode(T.self)
+            self.init(value: value)
+        }
+    }
+}
+
+public extension KeyedDecodingContainer {
+    func decodeIfPresent<T: ProtoEnum>(
+        _: T.Type,
+        forKey key: Key
+    ) throws -> T? {
+        guard contains(key) else {
+            return nil
+        }
+
+        let box = try decode(BoxedEnum<T>.self, forKey: key)
+        return box.value
+    }
+
+    func decode<T: ProtoEnum>(
+        _: Array<T>.Type,
+        forKey key: Key
+    ) throws -> [T] {
+        var container = try nestedUnkeyedContainer(forKey: key)
+
+        var results: [T] = []
+        if let count = container.count {
+            results.reserveCapacity(count)
+        }
+
+        while !container.isAtEnd {
+            let box = try container.decode(BoxedEnum<T>.self)
+            if let value = box.value {
+                results.append(value)
+            }
+        }
+
+        return results
+    }
+
+    func decode<T: ProtoEnum>(
+        _: Set<T>.Type,
+        forKey key: Key
+    ) throws -> Set<T> {
+        var container = try nestedUnkeyedContainer(forKey: key)
+
+        var results: Set<T> = []
+        if let count = container.count {
+            results.reserveCapacity(count)
+        }
+
+        while !container.isAtEnd {
+            let box = try container.decode(BoxedEnum<T>.self)
```

Comment:
(weird, I thought I already posted this comment)

I could decode `Array<BoxedEnum<T>>.self`, etc, but that would create an extra copy / iteration.
I'd prefer to just do manual iteration since this nominally benefits from a fast path.